lib/src/dependency.dart

```diff
@@ -200,7 +200,7 @@ class HostedDependency extends Dependency {
 
 @JsonSerializable(disallowUnrecognizedKeys: true)
 class HostedDetails {
-  final String name;
+  final String? name;
```

Comment:
Name doesn't really become optional... it just becomes inferred.

I'm not sure we can easily normalize this...

Reply:
I guess we'd need to "hoist" the value from the key above?

Or should we just drop name?

Reply:
It opens the question of whether or not we want `pubspec_parse` to really understand `pubspec.yaml`.
Currently, there is lots of subtleties it doesn't understand.

Reply:
The goal is to give a ~complete readonly object model for pubspec.yaml for tools that need to reason about it.

So "understand" is a bit subjective.

Reply:
If we want `HostedDetails` to represent what syntactically follows a `hosted: ` key in a pubspec, then IMO making `name` optional is the right thing to do because it really is optional now.  We don't have a `name` field on `PathDependency` or `GitDependency` either, even though one could argue that pub infers the value for those cases too.

There's also a hypothetical use-case in which a tool might want to distinguish between the `name` being duplicated in the `hosted` block, so just inferring may hide information (although we could add another field to distinguish those cases).

We could add another field (e.g. `String? declaredName`) and then make `name` a getter falling back to the inferred name. That way we stay close to pubspec semantics without breaking compatibility.

Reply:
I think 27de2a83efa0cc16e91fc348c78400baa53cb791 resolves this, PTAL.

Reply:
> So "understand" is a bit subjective.

For example, when interpreting `environment.sdk` (the SDK constraint), there is a lot of subtleties in `pub`. Like, if the lower-bound is less than `2.0.0-0`, then there is an implied upperbound of `< 2.0.0`, even if there is no upper-bound. And there is more corner cases like that.

Parsing a pubspec and interpreting a pubspec is not the same. And I'm not sure we want all the suble intepretation rules from `pub` in `pubspec_parse`, but maybe we do move more of that over time.

Reply:
On topic: given that `declaredName` allows us to remain backwards compatible, this seems like a good solution :D